students/MikeShand/lesson10/mailroom_fp_2.py

```diff
@@ -0,0 +1,245 @@
+            try:
+                donor_list[new_donor].add_donation(amount)
+                print('Thank you {} for your generous donation of ${:.2f}'.format(new_donor, amount))
+            except ValueError:
+                print("Please enter a round number!")
+        else:
+            new_donor_object = Donor(new_donor, [amount])
+            self.donor_list.append(new_donor_object)
+            print('Thank you {} for your generous donation of ${:.2f}'.format(new_donor, amount))
+
+    def donor_report(self):
+
+        '''Outputs a string that is a table of the donors and contributions'''
+
+        print('Here is a list of donors and contributions')
+        report = []
+        report.append('|{:<20}|{:<20}|{:<20}|{:<20}|'.format('Name', 'Total', 'Donations', 'Average'))
+        for donor in self.donor_list:
+           report.append('|{:<20}|{:>20}|{:>20}|{:>20}|'.format(donor.name, donor.total_donation(), donor.number_donations(), donor.avg_donation()))
+        return '\n'.join(report)
+
+    def print_donors(self):
+        print(self.donor_report())
+
+
+    def batch_file(self):
+
+        '''Creates a text file with a thank you message for each of the donors in the donor list'''
+
+        for donor_data in self.donor_list:
+            filename = donor_data.name.replace(" ", "_") + ".txt"
+            total_donation = donor_data.total_donation()
+            letter = ('Thank you {} for you generous contributions totalling {:.2f}!'.format(donor_data.name, total_donation))
+            with open(filename, 'w') as letter_file:
+                letter_file.write(letter)
+            print(f"{donor_data.name}'s letter has been saved to " + filename)
+
+    def challenge(self, factor, min=None, max=None):
+
+        '''Sends the current Donors through the map and filter function,
+           and returns a list of modified Donors'''
+
+        challenge = []
+        for name in self.donor_list:
+            challenge.append(Donor(name.name, self.map_filter(factor, name.donations, min, max)))
+        return challenge
+
+    def map_filter(self, factor, money, min=None, max=None):
+
+        '''Takes a map factor, a list of donations, and optional min and max donations,
+            and returns a list of modified donations'''
+
+        if min and max:
+            if min > max:
+                raise ValueError('Max must be greater than min')
+            else:
+                for donor in money:
+                    return list(map(lambda x: x * factor, filter(lambda y: min <= y <= max, money)))
+
+
+        elif min:
+            for donor in money:
+                return list(map(lambda x: x * factor, filter(lambda y: y >= min, money)))
+
+
+        elif max:
+            for donor in money:
+                return list(map(lambda x: x * factor, filter(lambda y: y <= max, money)))
+
+
+        else:
+            for donor in money:
+                return list(map(lambda x: x * factor, money))

```

Comment:
you want to avoid using `min` and `max` as var names, those are built-in functions that are being overwritten here.